### PR TITLE
homework-15

### PR DESCRIPTION
**Pull request by Bloodstinger**

### CODE REVIEW

File: src/lecture/fifteen/Main.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        try {
+            Injector.injectDependency();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        System.out.println("Если хотите сделать ставку, введите " +
+                "1.\n" +
```
Comment (Sofasmile) [inline src/lecture/fifteen/Main.java]: better write it in one line
 System.out.println("Если хотите сделать ставку, введите 1.\n"
and better move '+' at new line

File: src/lecture/fifteen/Main.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                "Если хотите внести автомобиль, введите " +
+                "2.");
+        Scanner scanner = new Scanner(System.in);
+        while (true) {
+            int command = scanner.nextInt();
+            if (command == 1) {
```
Comment (Sofasmile) [inline src/lecture/fifteen/Main.java]: try to use switch-case

File: src/lecture/fifteen/Main.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                    BetConsoleHandler.handle();
+                }
+            } else if (command == 2) {
+                System.out.println("Введите марку автомобиля и год выпуска");
+                try {
+                    break;
+                } finally {
+                    CarConsoleHandler.handle();
+                }
+            } else throw new IllegalArgumentException("Неверный аргумент");
```
Comment (Sofasmile) [inline src/lecture/fifteen/Main.java]: https://mate-academy.github.io/style-guides/java/java.html#s4.1.1-braces-always-used

File: src/lecture/fifteen/dao/BetDao.java [hunk 4]
```diff
@@ -0,0 +1,4 @@
+package lecture.fifteen.dao;
+
+public interface BetDao<T> extends DaoPattern<T> {
```
Comment (Sofasmile) [inline src/lecture/fifteen/dao/BetDao.java:3]: do we need this interface here?

File: src/lecture/fifteen/dao/BetDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    @Override
+    public void add(Bet bet) {
+        Storage.bets.add(bet);
+    }
+
+    @Override
+    public List<Bet> getAll() {
+        return Storage.bets;
+    }
+
```
Comment (Sofasmile) [inline src/lecture/fifteen/dao/BetDaoImpl.java]: please, remove redundant line everywhere

File: src/lecture/fifteen/dao/CarDao.java [hunk 6]
```diff
@@ -0,0 +1,4 @@
+package lecture.fifteen.dao;
+
+public interface CarDao<T> extends DaoPattern<T> {
```
Comment (Sofasmile) [inline src/lecture/fifteen/dao/CarDao.java:3]: do we need this interface?

File: src/lecture/fifteen/dao/DaoPattern.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package lecture.fifteen.dao;
+
+import java.util.List;
+
+public interface DaoPattern<T> {
```
Comment (Sofasmile) [inline src/lecture/fifteen/dao/DaoPattern.java:5]: better rename to Dao

File: src/lecture/fifteen/db/Storage.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package lecture.fifteen.db;
+
+import lecture.fifteen.model.Bet;
+import lecture.fifteen.model.Car;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Storage {
+    public static final List<Bet> bets = new ArrayList<>();
```
Comment (Sofasmile) [inline src/lecture/fifteen/db/Storage.java]:
```suggestion
    public static final List<Bet> BETS = new ArrayList<>();
```

File: src/lecture/fifteen/db/Storage.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+    public static final List<Car> cars = new ArrayList<>();
```
Comment (Sofasmile) [inline src/lecture/fifteen/db/Storage.java]:
```suggestion
    public static final List<Car> CARS = new ArrayList<>();
```

File: src/lecture/fifteen/factory/AbstractDaoFactory.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package lecture.fifteen.factory;
+
+import lecture.fifteen.dao.CarDaoImpl;
+import lecture.fifteen.dao.DaoPattern;
+
+public abstract class AbstractDaoFactory {
```
Comment (Sofasmile) [inline src/lecture/fifteen/factory/AbstractDaoFactory.java]: do we need this class?

File: src/lecture/fifteen/lib/ClassLoadHelper.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
+package lecture.fifteen.lib;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+
+public class ClassLoadHelper extends java.lang.ClassLoader {
```
Comment (Sofasmile) [inline src/lecture/fifteen/lib/ClassLoadHelper.java]:
```suggestion
public class ClassLoadHelper extends ClassLoader {
```

Comment (Sofasmile) [reply src/lecture/fifteen/lib/ClassLoadHelper.java]: do we need this class?

File: src/lecture/fifteen/lib/Injector.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    public static void injectDependency() throws IllegalAccessException {
+        Class<BetDaoImpl> betDaoClass = BetDaoImpl.class;
+        Class<CarDaoImpl> carDaoClass = CarDaoImpl.class;
+        Class<CarConsoleHandler> carConsoleHandlerClass = CarConsoleHandler.class;
+        Class<BetConsoleHandler> betConsoleHandlerClass = BetConsoleHandler.class;
+        Field[] betContollerFields = betConsoleHandlerClass.getDeclaredFields();
+        for (Field field : betContollerFields) {
+            if (field.getDeclaredAnnotation(Inject.class) != null) {
+                field.setAccessible(true);
+                if (field.getName().equals("betDao") && betDaoClass.getAnnotation(Dao.class) != null) {
```
Comment (Sofasmile) [inline src/lecture/fifteen/lib/Injector.java]: maybe better use equals by type than by name

File: src/lecture/fifteen/lib/Injector.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+                    field.set(null, BetDaoFactory.getInstance());
+                }
+            }
+        }
+        Field[] carControllerFields = carConsoleHandlerClass.getDeclaredFields();
+        for (Field field : carControllerFields) {
+            if (field.getDeclaredAnnotation(Inject.class) != null) {
+                field.setAccessible(true);
+                if (field.getName().equals("carDao") && carDaoClass.getAnnotation(Dao.class) != null) {
```
Comment (Sofasmile) [inline src/lecture/fifteen/lib/Injector.java]: the same

File: src/lecture/fifteen/dao/BetDao.java (hunk 4 above)
Comment (Bloodstinger) [reply src/lecture/fifteen/dao/BetDao.java:3]: Yes, that's an marker-interface, which is inherited from "DaoPattern" interface

File: src/lecture/fifteen/dao/CarDao.java (hunk 6 above)
Comment (Bloodstinger) [reply src/lecture/fifteen/dao/CarDao.java:3]: Same as "BetDao" interface.

File: src/lecture/fifteen/dao/DaoPattern.java (hunk 7 above)
Comment (Bloodstinger) [reply src/lecture/fifteen/dao/DaoPattern.java:5]: We already have Dao annotation in "lib", that would be confusing.

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**